packages/features/ee/billing/teams/internal-team-billing.ts

```diff
@@ -0,0 +1,153 @@
+import type { Prisma } from "@prisma/client";
+import type { z } from "zod";
+
+import { getRequestedSlugError } from "@calcom/app-store/stripepayment/lib/team-billing";
+import { purchaseTeamOrOrgSubscription } from "@calcom/features/ee/teams/lib/payments";
+import { MINIMUM_NUMBER_OF_ORG_SEATS, WEBAPP_URL } from "@calcom/lib/constants";
+import { getMetadataHelpers } from "@calcom/lib/getMetadataHelpers";
+import logger from "@calcom/lib/logger";
+import { Redirect } from "@calcom/lib/redirect";
+import prisma from "@calcom/prisma";
+import { teamMetadataSchema } from "@calcom/prisma/zod-utils";
+
+import billing from "..";
+import type { TeamBilling, TeamBillingInput } from "./team-billing";
+
+const log = logger.getSubLogger({ prefix: ["TeamBilling"] });
+
+const teamPaymentMetadataSchema = teamMetadataSchema.unwrap();
+
+export class InternalTeamBilling implements TeamBilling {
+  private _team!: Omit<TeamBillingInput, "metadata"> & {
+    metadata: NonNullable<z.infer<typeof teamPaymentMetadataSchema>>;
+  };
+  constructor(team: TeamBillingInput) {
+    this.team = team;
+  }
+  set team(team: TeamBillingInput) {
+    const metadata = teamPaymentMetadataSchema.parse(team.metadata || {});
+    this._team = { ...team, metadata };
+  }
+  get team(): typeof this._team {
+    return this._team;
+  }
+  private async getOrgIfNeeded() {
+    if (!this.team.parentId) return;
+    const parentTeam = await prisma.team.findUniqueOrThrow({
+      where: { id: this.team.parentId },
+      select: { metadata: true, id: true, parentId: true, isOrganization: true },
+    });
+    this.team = parentTeam;
+  }
+  private logErrorFromUnknown(error: unknown) {
+    let message = "Unknown error on InternalTeamBilling.";
+    if (error instanceof Error) message = error.message;
+    log.error(message);
+  }
+  async cancel() {
+    try {
+      const { subscriptionId } = this.team.metadata;
+      log.info(`Cancelling subscription ${subscriptionId} for team ${this.team.id}`);
+      if (!subscriptionId) throw Error("missing subscriptionId");
+      await billing.handleSubscriptionCancel(subscriptionId);
+      await this.downgrade();
+      log.info(`Cancelled subscription ${subscriptionId} for team ${this.team.id}`);
+    } catch (error) {
+      this.logErrorFromUnknown(error);
+    }
+  }
+  // New teams are published on creation, this is for backwards compatibility
+  async publish() {
+    const { url } = await this.checkIfTeamPaymentRequired();
+    const teamId = this.team.id;
+    if (url) throw new Redirect(307, url);
```

Comment:
This seems too far down in the core of the application to be handling Redirects. We should instead threw a specific application/core error and let the upper HTTP layers of the app handle the redirect.

Reply:
I'll see what i can come up with

Reply:
@sean-brydon and I talked and we have some ideas on how to "pull this up" a little bit.

Another thing to note is this code is reused by the API so having the inner core code trigger redirects could have weird side effects there.

Reply:
Addressed these changes here: https://github.com/calcom/cal.com/pull/15366/commits/dc1fcca00c77b5577802d4b75b121bd143a4fdeb 

Confirmed all places that use this handler automatically redirect on any data.url prop returned on the trpc `publish` mutation